### PR TITLE
First order logic working needs testing

### PR DESCRIPTION
**Pull request by daniely211**

### CODE REVIEW

File: src/reducers/index.js [hunk 1]
```diff
@@ -5,17 +5,20 @@ import {
   CHANGE_SYMBOL,
   UPDATE_RULE,
   ADD_CONSTANTS,
+  ADD_RELATIONS,
   ADD_STEP_DEPENDENCY,
   REMOVE_STEP_DEPENDENCY,
   UPDATE_STEP_DEPENDENCY,
   SET_STEP_DEPENDENCY
 } from '../constants'
+import _ from 'lodash'
```
Comment (jasonlipo) [inline src/reducers/index.js]: We've already added underscore.js
Change to `import _ from 'underscore'`

File: src/reducers/index.js [hunk 2]
```diff
@@ -62,6 +65,11 @@ const reducer = (state = initialState, action) => {
         newState.constants = [...new Set(newConstants)]
         return newState
 
+      case ADD_RELATIONS:
+        const newRelations = newState.relations.concat(action.payload)
+        newState.relations = _.uniqWith([...newRelations], _.isEqual)
```
Comment (jasonlipo) [inline src/reducers/index.js]: With underscore.js, you can just do `_.uniq(newRelations)`

File: src/reducers/index.js [hunk 3]
```diff
@@ -62,6 +65,11 @@ const reducer = (state = initialState, action) => {
         newState.constants = [...new Set(newConstants)]
```
Comment (jasonlipo) [inline src/reducers/index.js]: Might as well also use `_.uniq` here as well for consistency

File: src/translator/z3.js [hunk 4]
```diff
@@ -19,10 +19,11 @@ const declare_constants = (constants, file_contents) => {
 const declare_relations = (relations, file_contents) => {
   file_contents += '(declare-sort Type)\n'
   relations.forEach(rel => {
-    file_contents += 'declare-fun ' + rel.name + ' ('
-    rel.vars.forEach(() => {
+    file_contents += '(declare-fun ' + rel.name + ' ('
+    let i
+    for (i=0; i<rel.numParam; i++) {
```
Comment (jasonlipo) [inline src/translator/z3.js]: You can just do `for (let i=0; i<rel.numParam; i++)` without the line above

File: package.json [hunk 5]
```diff
@@ -18,6 +18,7 @@
     "fs": "0.0.1-security",
     "http": "^0.0.0",
     "interactjs": "^1.3.4",
+    "lodash": "^4.17.11",
```
Comment (jasonlipo) [inline package.json]: Use underscore.js, don't need this

File: src/parser/Propositional.g4 [hunk 6]
```diff
@@ -2,7 +2,7 @@ grammar Propositional;
 
 statement: expression;
 
-parameter:
+  parameter:
```
Comment (jasonlipo) [inline src/parser/Propositional.g4]: ???

File: src/components/index.js [hunk 7]
```diff
@@ -20,11 +20,11 @@ class IProve extends Component {
     }
   }
 
-  callZ3(steps, constants, i) {
+  callZ3(steps, constants, relations, i) {
     fetch('/z3', {
       method: "POST",
       headers: {"Content-Type": "application/json; charset=utf-8"},
-      body: JSON.stringify({ steps, constants, relations: [] })
+      body: JSON.stringify({ steps, constants, relations, i })
```
Comment (jasonlipo) [inline src/components/index.js]: I don't think we need to pass `i` to z3

File: src/reducers/index.js (hunk 2 above)
Comment (tomapopov) [reply src/reducers/index.js]: That didn't seem to work, so we used uniqWith

File: src/components/index.js [hunk 8]
```diff
@@ -52,7 +52,13 @@ class IProve extends Component {
           }
         })
         requiredSteps.push(step.ast)
-        this.callZ3(requiredSteps, constants, i)
+        this.callZ3(requiredSteps, constants, relations, i)
```
Comment (jasonlipo) [inline src/components/index.js]: Remove `i`

File: src/reducers/index.js (hunk 2 above)
Comment (daniely211) [reply src/reducers/index.js]: getting : Uncaught (in promise) ReferenceError: _ is not defined for _.iteratee('name')
 after i changed it to the line you recommend

File: src/components/index.js [hunk 9]
```diff
@@ -52,7 +52,13 @@ class IProve extends Component {
...
+      } else {
+       // if there are no dependencies change the box to red
+       if (step.type !== "goal") {
+          // this.callZ3(step.ast, constants, relations, i)
```
Comment (jasonlipo) [inline src/components/index.js]: Remove

Comment (daniely211) [reply src/components/index.js]: oh yes sorry that was us trying to fiddle with the boxes when it didnt have any dependencies it should just pass it to z3 that line... but we didnt know how to

File: src/parser/Propositional.g4 (hunk 6 above)
Comment (daniely211) [reply src/parser/Propositional.g4]: !!!

**Pull request merged**